File: src/azure_region_monitor/assets/reader-feedback.js
(() => {
  "use strict";
  const contextNode = document.getElementById("reader-context");
  if (!contextNode) return;
  const context = JSON.parse(contextNode.textContent);
  const get = id => document.getElementById(`reader-${id}`);
  const role = get("role");
  const prior = get("prior");
  const start = get("start");
  const skip = get("skip");
  const status = get("status");
  const reading = get("reading");
  const answers = get("answers");
  const response = get("response");
  const exported = get("export");
  const budget = context.reading_budget_ms;
  let state = "ready";
  let started = null;
  let startedAt = null;
  let readingElapsed = null;
  let readingInterrupted = false;
  let answeringInterrupted = false;
  let endReason = null;
  let timeout = null;
  let clock = null;
  let attemptId = null;
  let preparedPacket = null;

  function markdownText(value) {
    return String(value ?? "Not answered").replace(/&/g, "&amp;")
      .replace(/</g, "&lt;").replace(/>/g, "&gt;")
      .replace(/([\\`*_[\]{}()#!|~])/g, "\\$1");
  }

  function draftSummary(packet) {
    const yesNo = value => value ? "yes" : "no";
    const milliseconds = value => value === null ? "not timed" : `${value} ms`;
    const answer = (question, value) =>
      `### ${question}\n> ${markdownText(value || "Not answered").replace(/\r?\n/g, "\n> ")}`;
    const proofLabels = {
      deployment: "The listed features were successfully deployed",
      observations: "Only the catalog or measurement observations stated",
      capacity: "Capacity and quota are available for my subscription",
      retirement: "Missing listings are confirmed retirements",
      unsure: "Not sure",
    };
    const choiceLabels = {yes: "Yes", no: "No", unsure: "Not sure"};
    const result = packet.answers;
    return [
      "## Optional reading check",
      "Eligibility below is not a correctness score. Founder/repeat-reader feedback is qualitative.",
      `- Date: ${markdownText(packet.date)}; protocol: ${markdownText(packet.protocol)}`,
      `- Case ID: ${markdownText(packet.case_id)}; view ID: ${markdownText(packet.view_id)}`,
      `- Attempt ID: ${markdownText(packet.attempt_id)}`,
      `- Snapshots: ${markdownText(packet.previous_timestamp ?? "none")} → ${markdownText(packet.current_timestamp)}`,
      `- Method: ${packet.method}; role: ${markdownText(packet.role)}; prior exposure: ${markdownText(packet.prior_exposure)}`,
      `- Started (UTC): ${markdownText(packet.started_at)}; viewport class: ${packet.viewport}`,
      `- Reading exposure: ${milliseconds(packet.reading_elapsed_ms)}; budget: ${packet.reading_budget_ms} ms; overrun: ${milliseconds(packet.timer_overrun_ms)}`,
      `- End reason: ${packet.end_reason}; interrupted reading/answering: ${yesNo(packet.reading_interrupted)}/${yesNo(packet.answering_interrupted)}; help/reopening: ${yesNo(packet.used_help)}`,
      `- Eligibility suggestion: ${packet.candidate_for_scoring ? "candidate for human comprehension review" : "qualitative/repeat-exposure only"}`,
      answer("What changed most?", result.main_change),
      answer("Affected region", result.affected_region),
      answer("What does the evidence establish?", proofLabels[result.evidence_meaning]),
      answer("Do zero new delistings prove earlier recovery?", choiceLabels[result.zero_delistings_means_recovery]),
      answer("Feature and distinction", result.feature_distinction),
      answer("Clarity", result.clarity ? `${result.clarity}/5` : null),
      answer("What was confusing or missing?", result.confusion),
      "Full JSON record: optionally attach the reviewed download manually. Nothing is submitted by the dashboard.",
    ].join("\n\n");
  }

  function updateStart() {
    start.disabled = skip.disabled = state !== "ready" || !role.value || !prior.value;
  }
  role.addEventListener("change", updateStart);
  prior.addEventListener("change", updateStart);
  updateStart();

  function begin(timed) {
    if (state !== "ready" || !role.value || !prior.value) return;
    attemptId = crypto.randomUUID();
    role.disabled = prior.disabled = start.disabled = skip.disabled = true;
    startedAt = new Date().toISOString();
    if (!timed) {
      state = "answering";
      endReason = "skipped";
      response.querySelectorAll("[required]").forEach(input => { input.required = false; });
      answers.hidden = false;
      status.textContent = "Qualitative feedback only; no timed reading measurement.";
      get("main").focus();
      return;
    }
    state = "reading";
    started = performance.now();
    reading.hidden = false;
    status.textContent = "Read the briefing. It will hide after 15 seconds, or when you choose to answer.";
    reading.scrollIntoView({block: "start"});
    clock = setInterval(() => {
      const remaining = Math.max(0, Math.ceil((budget - (performance.now() - started)) / 1000));
      get("clock").textContent = `${remaining} seconds remaining`;
    }, 100);
    timeout = setTimeout(() => finish("budget"), budget);
    get("ready").focus({preventScroll: true});
  }

  function finish(reason) {
    if (state !== "reading") return;
    readingElapsed = Math.round(performance.now() - started);
    endReason = reason;
    clearTimeout(timeout);
    clearInterval(clock);
    reading.hidden = true;
    state = "answering";
    answers.hidden = false;
    status.textContent = "Reading finished. Answer from memory; no feedback has been sent.";
    get("main").focus();
  }

  start.addEventListener("click", () => begin(true));
  skip.addEventListener("click", () => begin(false));
  get("ready").addEventListener("click", () => finish("ready"));
  document.addEventListener("visibilitychange", () => {
    if (!document.hidden) return;
    if (state === "reading") {
      readingInterrupted = true;
      finish("interrupted");
    } else if (state === "answering") {
      answeringInterrupted = true;
    }
  });

  response.addEventListener("input", () => {
    if (state === "prepared") state = "answering";
    preparedPacket = null;
    exported.hidden = true;
    get("draft").href = "#reader-export";
    get("draft-status").textContent = "";
  });
  response.addEventListener("submit", event => {
    event.preventDefault();
    if (!["answering", "prepared"].includes(state) || !response.reportValidity()) return;
    const helped = get("helped").checked;
    const overrun = readingElapsed === null ? null : Math.max(0, readingElapsed - budget);
    const candidate = endReason !== "skipped" && prior.value === "no" &&
      !readingInterrupted && !answeringInterrupted && !helped && overrun <= 250;
    const packet = {
      ...context,
      attempt_id: attemptId,
      started_at: startedAt,
      method: endReason === "skipped" ? "qualitative" : "timed",
      role: role.value,
      prior_exposure: prior.value,
      viewport: innerWidth < 720 ? "small" : "large",
      reading_elapsed_ms: readingElapsed,
      timer_overrun_ms: overrun,
      end_reason: endReason,
      reading_interrupted: readingInterrupted,
      answering_interrupted: answeringInterrupted,
      used_help: helped,
      candidate_for_scoring: candidate,
      answers: {
        main_change: get("main").value.trim(),
        affected_region: get("region").value.trim(),
        evidence_meaning: get("proof").value || null,
        zero_delistings_means_recovery: get("recovery").value || null,
        feature_distinction: get("feature").value.trim(),
        clarity: get("clarity").value ? Number(get("clarity").value) : null,
        confusion: get("confusion").value.trim(),
      },
    };
    const encoded = JSON.stringify(packet);
    preparedPacket = packet;
    get("packet").value = encoded;
    const timing = readingElapsed === null ? "No reading timer was used. " :
      `Recorded reading exposure: ${(readingElapsed / 1000).toFixed(2)} seconds. `;
    const reasons = [];
    if (prior.value !== "no") reasons.push("prior exposure is not confirmed as first-time");
    if (readingInterrupted || answeringInterrupted) reasons.push("the tab was interrupted");
    if (helped) reasons.push("help or reopening was reported");
    if (overrun > 250) reasons.push("timer delivery exceeded the permitted overrun");
    get("result-note").textContent = timing + (candidate ?
      "This first-exposure attempt can be reviewed for comprehension. This is not a correctness score." :
      `Keep this as qualitative/repeat-exposure feedback${reasons.length ? ": " + reasons.join("; ") : ""}, separate from first-exposure timed results.`);
    get("copy-status").textContent = "";
    get("draft").href = "#reader-export";
    get("draft-status").textContent = "";
    state = "prepared";
    exported.hidden = false;
    get("packet").focus();
  });

  get("draft").addEventListener("click", event => {
    try {
      if (!preparedPacket) throw new Error("Prepare your local result first.");
      if (!window.AzureMonitorFeedback?.issueUrl) {
        throw new Error("The GitHub draft helper is unavailable.");
      }
      get("draft").href = window.AzureMonitorFeedback.issueUrl(
        `Optional reading check: ${preparedPacket.date}`,
        draftSummary(preparedPacket),
        context.repository_url,
      );
      get("draft-status").textContent = "Opening a reviewable GitHub draft, not submitting an issue.";
    } catch (error) {
      event.preventDefault();
      get("draft").href = "#reader-export";
      get("draft-status").textContent = `${error.message} Your full local result and JSON download are still available. Nothing was truncated or submitted.`;
    }
  });

  get("copy").addEventListener("click", async () => {
    try {
      await navigator.clipboard.writeText(get("packet").value);
      get("copy-status").textContent = "Copied the full local JSON record. Nothing was submitted; review it before sharing.";
    } catch (error) {
      get("packet").focus();
      get("packet").select();
      get("copy-status").textContent = `Clipboard unavailable (${error.name}). Copy the selected packet manually.`;
    }
  });
  get("download").addEventListener("click", () => {
    const blob = new Blob([get("packet").value + "\n"], {type: "application/json"});
    const url = URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = url;
    link.download = `reader-feedback-${context.date}-${attemptId}.json`;
    link.click();
    setTimeout(() => URL.revokeObjectURL(url), 1000);
    get("copy-status").textContent = "A local JSON download was requested. Nothing was submitted.";
  });
})();

File: src/azure_region_monitor/reader_feedback.py
"""Optional, browser-local comprehension checks with a reviewable GitHub draft."""

from __future__ import annotations

import html
import json
from typing import Any

from azure_region_monitor.briefing_view import has_briefing, render_briefing
from azure_region_monitor.feedback_context import measurement_context as measurement_context
from azure_region_monitor.feedback_context import presentation_id as presentation_id


def render_reading_check_page(day: dict[str, Any] | None, style: str, view_id: str) -> str:
    title = "Optional briefing reading check"
    if day is None or not has_briefing(day):
        body = '<p>A complete daily briefing is not available for a reading check yet. Please try again after the next snapshot comparison.</p>'
    else:
        context = measurement_context(day, view_id)
        encoded = json.dumps(context, ensure_ascii=True).replace("<", "\\u003c")
        body = f"""
        <section class="panel feedback-intro">
          <h2>A voluntary 15-second reading check</h2>
          <p>This optional research check is not the main feedback path.
             For maintainer observations, use <a href="/feedback.html">page feedback</a> instead.
             Founder or repeat-reader observations are qualitative, not unbiased first-time-reader results.</p>
          <p>Read the briefing, then answer from memory: what changed most, where, and what does the evidence prove?
             This measures comprehension, not whether you liked the design.</p>
          <p><strong>Privacy:</strong> answers and timing stay in this page's memory.
             No analytics, cookies, background submissions, or browser storage.
             Ordinary page/evidence requests still reach the static host.
             Opening a GitHub draft sends its summary to GitHub, but does not submit an issue.
             You review it and choose the final Submit on GitHub.</p>
          <p><strong>Public feedback:</strong> submitted issues and attachments in this repository are public.
             Do not share secrets or private information. The dashboard needs no backend or authentication token.</p>
          <p>Use one attempt. Switching away from this tab marks it interrupted; it does not silently pause the clock.
             You may stop or skip at any time. Do not enter names, email addresses, subscription IDs, or private workload details.</p>
          <p>No screenshot capture is offered in this timed check. Use normal page feedback for optional screenshots.</p>
          <div class="feedback-profile">
            <label for="reader-role">Your perspective<select id="reader-role" required>
              <option value="">Choose a perspective</option>
              <option value="architect">Cloud architect</option><option value="sre">SRE / operations</option>
              <option value="manager">Manager</option><option value="engineer">Engineer / technical reader</option>
              <option value="other">Another perspective / prefer not to say</option>
            </select></label>
            <label for="reader-prior">Have you already read this dated briefing?<select id="reader-prior" required>
              <option value="">Choose an answer</option><option value="no">No</option>
              <option value="yes">Yes</option><option value="unsure">Not sure</option>
            </select></label>
          </div>
          <p>Case: {html.escape(day['date'])}. The timer limits reading, not answering.
             This is a self-selected check, not a controlled experiment.</p>
          <button type="button" id="reader-start" disabled>Start reading</button>
          <button type="button" id="reader-skip" disabled>Skip timer; leave qualitative feedback</button>
          <p role="status" id="reader-status">Choose your perspective and prior familiarity to begin.</p>
          <noscript><p>JavaScript is needed for the local check. You can still read the dashboard.</p></noscript>
        </section>
        <section id="reader-reading" hidden aria-label="Timed reading">
          <div class="feedback-timer"><strong id="reader-clock" aria-live="off">15 seconds</strong>
            <button type="button" id="reader-ready">I'm ready to answer</button></div>
          {render_briefing(day, include_feedback=False)}
        </section>
        <section class="panel feedback-answers" id="reader-answers" hidden>
          <h2>Answer without reopening the briefing</h2>
          <form id="reader-response">
            <label for="reader-main">1. What changed most since the previous scan?
              <textarea id="reader-main" rows="3" maxlength="500" required></textarea></label>
            <label for="reader-region">2. Name one region affected by a new change or evidence gap. If there were none, say so.
              <input id="reader-region" maxlength="200" required></label>
            <label for="reader-proof">3. What does this briefing establish?
              <select id="reader-proof" required><option value="">Choose an answer</option>
                <option value="deployment">The listed features were successfully deployed</option>
                <option value="observations">Only the catalog or measurement observations stated</option>
                <option value="capacity">Capacity and quota are available for my subscription</option>
                <option value="retirement">Missing listings are confirmed retirements</option>
                <option value="unsure">Not sure</option></select></label>
            <label for="reader-recovery">4. Does zero NEW delistings prove that previously missing listings recovered?
              <select id="reader-recovery" required><option value="">Choose an answer</option>
                <option value="yes">Yes</option><option value="no">No</option><option value="unsure">Not sure</option>
              </select></label>
            <label for="reader-feature">5. Name one feature and explain what distinguishes it. (Optional)
              <textarea id="reader-feature" rows="3" maxlength="400"></textarea></label>
            <label for="reader-clarity">6. How easy was the briefing to understand?
              <select id="reader-clarity" required><option value="">Choose a rating</option>
                <option value="1">1 - Very difficult</option><option value="2">2 - Difficult</option>
                <option value="3">3 - Mixed</option><option value="4">4 - Easy</option><option value="5">5 - Very easy</option>
              </select></label>
            <label for="reader-confusion">7. What was confusing, missing, or hard to find? (Optional)
              <textarea id="reader-confusion" rows="3" maxlength="500"></textarea></label>
            <label class="feedback-helped"><input id="reader-helped" type="checkbox">
              I used another tab, reopened the briefing, or received help while answering.</label>
            <button type="submit">Prepare my result locally</button>
          </form>
        </section>
        <section class="panel feedback-export" id="reader-export" hidden>
          <h2>Review before sharing</h2>
          <p id="reader-result-note"></p>
          <p>This is <strong>not submitted</strong>. Open a GitHub draft to review a human-readable summary
             of your answers, case/view IDs, timing, and eligibility. No raw JSON packet goes in the URL.
             Only your final Submit on GitHub creates the public issue.</p>
          <p>The full local record includes your answers, timing, role, case ID, and presentation version,
             with no expected answers or automatic grade. You may download JSON and manually attach it
             to the draft. Review it first: attachments upload immediately when pasted or attached on GitHub,
             even before you submit the issue.</p>
          <label for="reader-packet">Your full local record (JSON)<textarea id="reader-packet" rows="9" readonly></textarea></label>
          <div class="feedback-actions"><button type="button" id="reader-copy">Copy packet</button>
            <button type="button" id="reader-download">Download JSON</button></div>
          <p role="status" id="reader-copy-status"></p>
          <a id="reader-draft" href="#reader-export" target="_blank" rel="noopener noreferrer"
             referrerpolicy="no-referrer">Open GitHub draft</a>
          <p id="reader-draft-status" role="alert"></p>
          <p>GitHub may log the draft URL and requires its own sign-in to submit. The dashboard never asks for a token.
             If the summary exceeds the draft URL limit, the full local result and JSON download remain available;
             nothing is truncated or automatically sent.</p>
        </section>
        <script id="reader-context" type="application/json">{encoded}</script>
        <script src="/assets/github-feedback.js" defer></script>
        <script src="/assets/reader-feedback.js" defer></script>"""
    return f"""<!doctype html><html lang="en"><head>
      <meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1">
      <meta name="robots" content="noindex"><meta name="referrer" content="no-referrer">
      <title>{title}</title>{style}</head><body><main class="content-page reader-study">
      <header><h1>{title}</h1><a href="/">Back to dashboard</a>
        · <a href="/feedback.html">Maintainer page feedback</a></header>{body}
      </main></body></html>"""
